feat:  Agent Config Version Manager

mindhash opened the pull request:
This PR introduces:
- Agent Config versioning and deployment 
- Otel Config Prep in opAmp: Ingestion Rules, lb exporter 
- Support for handling traces in multiple agents setup through lb exporter

srikanthccv commented on pkg/query-service/app/opamp/pipeline_builder.go:
```diff
@@ -0,0 +1,196 @@
+	inserts := 0
+
+	// go through plan again in the increasing order
+	for i := 0; i < len(spec); i++ {
```
I am trying to understand this logic. Let's say I have ['spanmetrics', 'attribute', 'batch'] in my original config. And now, the user wants to enable the sampling, IIUC, this adds the 'signoz_tail_sampling' in the first position, which will affect the span metrics. How is this situation handled generally in this implementation?

mindhash replied:
The overrides can happen at 2 places.  In our config (will get to this later) and directly in config file. 

In this case,  user can add `signoz_tail_sampling` manually in the list to enforce an order.   If we find signoz_tail_sampling we would not add it again or change its position. 

`processor: [spanmetrics, signoz_tail_sampling]  `

If config file has the above order, it will be maintained.

This does require some documentation for user but addresses the problem of coordination.

mindhash replied:
Regarding some of the dependencies that we wish to enforce, we can also use tracePipelinePlan{} to ensure a correct order.  All the processors in this map need not be custom - for e.g. signoz_tail_sampling.

srikanthccv replied:
> user can add signoz_tail_sampling manually in the list to enforce an order

Why does user need to intervene manually? For instance, they didn't have the `signoz_tail_sampling` earlier and they are configuring it from UI. In this case it adds in order which results in incorrect span metrics.

>Regarding some of the dependencies that we wish to enforce, we can also use tracePipelinePlan{} to ensure a correct order.

Take the above example and what does this look like?

mindhash replied:
> Why does user need to intervene manually? For instance, they didn't have the signoz_tail_sampling earlier and they are configuring it from UI. In this case it adds in order which results in incorrect span metrics.

Since the config editing is shared between UI (signoz) and the user (directly in the file). some level of co-ordination would be required. What we have implemented so far, covers basic use case where we don't overwrite what is already there.  

The current assumption is user would not have dependencies on our order and we don't on user edits.  But if there is such a scenario, we would have to co-ordinate.   IMO, as long as we can explain how we edit config files, users should be able to make their edits.  

In case you have suggestions for better way to handle co-ordinated edits with user, please create a separate issue if thats ok.

srikanthccv replied:
The discussion has gone in a different direction. Let me bring up my original point; this logic adds a tail sampling processor before the spanmetrics and spanmetrics is the most basic use case. Adding a tail sampling processor before span metrics is incorrect. How do we address it? I don't think the answer should be we communicate to users and expect it to resolve by them.

srikanthccv replied:
@mindhash will create a new issue to track this.

srikanthccv commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+// StartNewVersion launches a new config version for given set of elements
+func StartNewVersion(ctx context.Context, eleType ElementTypeDef, elementIds []string) (*ConfigVersion, error) {
+
+	if !m.Ready() {
+		// agent is already being updated, ask caller to wait and re-try after sometime
+		return nil, fmt.Errorf("agent updater is busy")
+	}
```
If the suggestion for caller is to wait and retry the I would suggest make channel and block here instead returning error.

mindhash replied:
The function would return ERROR to the caller.

srikanthccv commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+
+	configVersion, err := GetConfigVersion(ctx, typ, version)
+	if err != nil {
+		zap.S().Debugf("failed to fetch config version during redeploy", err)
```
Please use format specifiers. This spits out an ugly message such as `%!(EXTRA *errors.errorString=...`

mindhash replied:
got it. Done

srikanthccv commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+		return fmt.Errorf("failed to fetch details of the config version")
+	}
+
+	if configVersion == nil || (configVersion != nil && configVersion.LastConf == "") {
```
What makes a config valid & deployable? Can you move that logic to some `canDeploy` or some receiver on `ConfigVersion`?

mindhash replied:
can check the status of config version before pushing update. Also we are not tracking IsValid in configversion yet.  will add it.

srikanthccv commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+			return err
+		}
+
+		m.updateDeployStatus(ctx, ElementTypeSamplingRules, version, string(DeployInitiated), "Deployment started", configHash, configVersion.LastConf)
```
Same question here.

mindhash replied:
not sure. what is the question? this method updates status of version.
The version is created upfront before we push updates.

srikanthccv commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+		}
+
+		opamp.AddToMetricsPipelineSpec("filter")
+		configHash, err := opamp.UpsertControlProcessors(ctx, "metrics", processorConf, m.OnConfigUpdate)
```
When is version updated after change?

mindhash replied:
Version is updated prior to change. The calls (logs processor or ingestion controller) book a new version before pushing updates.

srikanthccv commented on pkg/query-service/agentConf/sqlite/init.go:
```diff
@@ -0,0 +1,65 @@
+		created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP, 
+		updated_by TEXT,
+		updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP, 
+		version INTEGER DEFAULT 1,
```
Should this be unique?

srikanthccv reviewed (commented):
I would have split this into parts 1. agent config management 2. lb & ingestion control 3. pipeline management. 

I reviewed the agent config part. Is this atleast tested with minimal changes once? I would appreciate some sort of documentation for deployment.

mindhash commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+			return fmt.Errorf("failed to deploy the config")
+		}
+
+		m.updateDeployStatus(ctx, ElementTypeSamplingRules, version, string(DeployInitiated), "Deployment started", configHash, configVersion.LastConf)
```
The lastConf would be used to redeploy an older config.  This would be particularly useful when a new agent (fresh config) collector joins the ring.  We can pull latest configs from all 3 element types, merge and deploy.  

hence, we keep record of what was pushed. This reduces the effort of having to rebuild config and also cycling-dependency of agentConf on calling packages (e.g. ingestionRules)

mindhash replied:
The agent config versions table stores versioned configs for each types - Logs, Sampling Rules and Drop rules. 

Together [agent_config_versions, agent_config_elements, log pipelines or rules] make up signoz definition of config.  These are used to prepare part of a collector config when the user is trying to push changes. 
 
The prepared collector part is stored in agent_config_versions table for each version. So in case we want to redeploy we don't fetch log pipelines and rules to re-make the collector config.

srikanthccv replied:
>The agent config versions table stores versioned configs for each types - Logs, Sampling Rules and Drop rules.

Yes, this was clear. I understood this.

>Together [agent_config_versions, agent_config_elements, log pipelines or rules] make up signoz definition of config. These are used to prepare part of a collector config when the user is trying to push changes.

Where is the "signoz definition of config" defined?

> So in case we want to redeploy we don't fetch log pipelines and rules to re-make the collector config.

What is `Redeploy` supposed to do? Does it push the config for an element type `typ` with version `version` to the collector? If the idea is to fetch config from DB and push, why does it call `opamp.AddToTracePipelineSpec`?


> This would be particularly useful when a new agent (fresh config) collector joins the ring. We can pull latest configs from all 3 element types, merge and deploy.

The `UpsertControlProcessors` goes over all agents and pushes the changes. When a new agent joins we know if it is newly created. Why update all the agents instead of a newly joined agent?

mindhash commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+
+// Ready indicates if Manager can accept new config update requests
+func (mgr *Manager) Ready() bool {
+	if mgr.lock != 0 {
```
This was misssed as i had used mutex before. Will add atomic.LoadUint64

Atomic is used to check if lock is acquired by other process.  Couldnt find a way to check lock in mutex

mindhash replied:
here we are using a singleton pattern to make config requests mutually exclusive. We are blocking users in the UI when a config update is already in progress. This is to reduce concurrency complications.  
I understand the approach is not native to go. but this was decided in earlier calls (if you remember) to avoid concurrency issues

mindhash commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,224 @@
+		return fmt.Errorf("config must have atleast one element")
+	}
+
+	if c.Version != 0 {
```
This is the struct 
ConfigVersion {
  Version int
} 

The versions in insertConfig need to be auto-assigned and monotonically increased. Hence if InsertConfig is called with a set version we reject the request. The version has to be unassigned. 
Could have used *int instead of int but the versions start with 1. So this is still workable.

mindhash replied:
Ok I will add a comment in the code to explain

mindhash commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,224 @@
+		deploy_result 
+		FROM agent_config_versions 
+		WHERE element_type = $1 
+		AND version = ( 
+			SELECT MAX(version) 
+			FROM agent_config_versions 
+			WHERE element_type=$2)`, typ, typ)
```
Not much diff there. I have an index on (version, element_type) to make MAX() faster in most cases. 
OrderBy does use index at times but not always guaranteed.
Again this depends on db specific optimzer.  But in my experience MAX() plays out pretty well

mindhash commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,224 @@
+	}
+
+	if c.Version != 0 {
+		zap.S().Errorf("invalid version assignment while inserting agent config", c.Version, c.ElementType)
```
will do

mindhash commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,224 @@
+	if err != nil {
+		if err != sql.ErrNoRows {
+			zap.S().Errorf("failed to fetch latest config version", err)
+			return fmt.Errorf("failed to fetch latest config version")
```
This is allows no-rows situation flow through. So We only return exception when GetLatestVersion returns error  that is other than no-rows. 

`		if err != sql.ErrNoRows {
`

mindhash commented:
@srikanthccv I have made following updates:

- reverse merge opamp with latest develop changes. had to change server.go
- removed zap.S().Errorf and replaces with zap.S().Error
- added comments for version!= 0 to describe it in detail 

I could not test remove ingestion rules (tail sampling) spec preparation code as it had multiple deps.   

Overall this PR is just contains supporting packages. A lot of testing has been done by through calling routine (ingestion rules api). I have a local branch that I use for testing and cherry-pick files for this branch.

nityanandagohain commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,227 @@
+	configQuery := `INSERT INTO agent_config_versions(	
+		id, 
+		version, 
+		element_type, 
```
The created_by value is not inserted here, can you please checkout my changes from here https://github.com/SigNoz/signoz/blob/94333d6152a09b8c2809ff735e00f532b973d7f5/pkg/query-service/agentConf/db.go

mindhash replied:
Added created_by

srikanthccv commented on pkg/query-service/app/opamp/configure_lb.go:
```diff
@@ -0,0 +1,277 @@
+		"processors": []string{},
+		"exporters":  []string{LbExporterName},
+	}
+	preUpdateTraces["receivers"] = []string{OtlpInternalReceiver}
```
Here is my understanding of this,

- this function receives the current effective config of agent
- a new internal receiver is prepared for OTLP
- a new lb exporter is prepared
- merge both of them with the current effective config
- at this point, this configuration has both the regular receivers at 4317/18 and internal receiver at 4949
- take the "receivers" from the "traces" pipeline

What is the strategy here? Does the same receiver act as an agent that uses the lb exporter and lb receiver?

mindhash replied:
here is the target config.  The LB exporter will send traces to otlp_internal.  I have described this in another comment

	//  traces/lb:
	// 		receivers: [otlp, jaeger]
	// 		processors: []
	// 		exporters: [lbExporter]

	// update receiver in service > pipelines > traces
	// traces:
	//		receivers: [otlp_internal]
	//		processors: [signoz_tail_sampling, batch]
	//		exporters: [clickhousetraceexporter]

srikanthccv commented on pkg/query-service/app/opamp/configure_ingestionRules.go:
```diff
@@ -0,0 +1,228 @@
+				},
+			},
+		},
+	}
```
This is hard to follow/maintain with all the `map[string]interface{}`. Let's use the config updater. It would have been if this was occasionally used, but it is everywhere.

mindhash replied:
Yes. I am going to replace it with actual params. 
I am not keen on getting config updater merged. It introduces unnecessary dependencies. It is going to be even harder to maintain.

srikanthccv commented on pkg/query-service/app/opamp/configure_ingestionRules.go:
```diff
@@ -0,0 +1,228 @@
+		return confHash, err
+	}
+
+	agent.SendToAgent(&protobufs.ServerToAgent{
```
I don't think this should send the status or interact with each agent directly. There is one already on the `Server` why this again?

mindhash replied:
do you mean onMessage()? I am not sure this was implemented by Nitya. Can you share which method to be used instead of this to interact with agent?

srikanthccv replied:
Not `onMessage`, the server also has `UpsertProcessor` function added in this PR which I believe is supposed to do the same.

mindhash replied:
UpsertProcessor() is being used for log pipelines which Nitya is working on. He wanted a different level of control on how processors are inserted.

For ingestion rules, I have UpsertIngestionControlProcs() and for log pipelines there is UpsertProcessor().
Both can be separate in this release as it helps us move in different direction (me and Nitya). We can merge them in future versions once this whole things becomes stable.

mindhash replied:
yeh. agree. I will include this in separate PR tomorrow.  Added issue for all improvements[here](https://github.com/SigNoz/engineering-pod/issues/963)

srikanthccv commented on pkg/query-service/app/opamp/configure_lb.go:
```diff
@@ -0,0 +1,277 @@
+
+// prepareNonLbAgentSpec creates non-LB exporter agents that handle otlp receiver
+// only.
+func prepareNonLbAgentSpec(agentConf *confmap.Conf) (serviceConf *confmap.Conf, fnerr error) {
```
Can I see this in action with three or four collectors running somewhere?

mindhash replied:
No yet. We will be working on multi-collector config setup and testing this week.

mindhash replied:
btw, the config preparation part is tested. But k8s level testing is not done yet

nityanandagohain commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+	message := "deploy successful"
+
+	defer func() {
+		zap.S().Error(status, zap.String("agentId", agentId), zap.String("agentResponse", message))
```
Should this be info instead of error as it is getting printed every time even if it's successful?

mindhash replied:
one

mindhash commented on pkg/query-service/app/opamp/configure_ingestionRules.go:
```diff
@@ -0,0 +1,228 @@
+
+		// check if lb pipeline is setup for lb agents, here we assume
+		// that otlp_internal will exist implicitly if traces/lb pipeline exists
+		if agent.CanLB && !checkPipelineExists(agentConf, TracesLbPipelineName) {
```
The agents will be made LB capable on the collector side.  
https://github.com/mindhash/signoz-otel-collector/blob/feat/lb-exporter/opamp/server_client.go#L97

When setting up k8s or dockerconfig, separate pools for LB and non-LB can be setup. The option is mainly for us to evaluate different configuration and not really for customer to play around. 

At the moment, I am not sure if we really need each collector to be LB  always. An ideal config would be one or two LB and rest of them receivers. 

Some of the multi-collector config is still not clear from k8s perspective. we may have to re-visit logic in this week.

mindhash commented on pkg/query-service/app/opamp/configure_ingestionRules.go:
```diff
@@ -0,0 +1,228 @@
+				zap.S().Error("failed to prepare OTLP internal receiver spec for agent", agent.ID, err)
+				return confHash, err
+			}
+			if err := agentConf.Merge(serviceConf); err != nil {
+				zap.S().Error("failed to merge OTLP internal receiver spec for agent", agent.ID, err)
+				return confHash, err
+			}
```
This is to support one collector with 2 pipelines 
- LB exporter (receiver listens on :4317)
- internal receiver (listens on :4949)

The pipeline with LB exporter will send traces to internal receiver (listening on port :4949). The above configuration sets up 2 pipelines in the same collector config.

mindhash commented on pkg/query-service/app/opamp/configure_lb.go:
```diff
@@ -0,0 +1,277 @@
+	receivers[OtlpInternalReceiver] = map[string]interface{}{
+		"protocols": otlpreceiver.Protocols{
+			GRPC: &otlpreceiver.GRPCServerSettings{
+				Endpoint: "0.0.0.0:4949",
+			},
+			HTTP: &otlpreceiver.HTTPServerSettings{
+				Endpoint:   "0.0.0.0:4949",
```
Multi-collector setup is not tested as it depends on k8s config. The current code works for single node installations.

mindhash replied:
removed this from PR. Will be submitting a separate PR for multi-collector after testing.

mindhash commented on pkg/query-service/app/opamp/configure_lb.go:
```diff
@@ -0,0 +1,277 @@
+		"protocol": map[string]interface{}{
+			"otlp": map[string]interface{}{
+				// Timeout is the timeout for every attempt to send data to the backend.
+				"timeout": "1s",
```
agree. will do.

mindhash commented on pkg/query-service/app/opamp/configure_lb.go:
```diff
@@ -0,0 +1,277 @@
+	// perform above in dry-run mode and then final mode
+	// if all agents succeed exit success else fail
+
+	receivers := agentConf.Get("receivers").(map[string]interface{})
```
It allows flexibility in handling config and keeping up with upstream config. I am not sure config updater should be used to introduce the collector dependencies.   
By adding collector dep, we have to keep up the libs in sync with every dependency the collector and its deps have.  I tried doing this and I would suggest keeping this code (even if it is not pretty) vs adding collector dep. 

cc: @nityanandagohain @ankitnayan  in case you have any thoughts here, do share

mindhash replied:
I added ConfigParser{} to make the improve readability and maintainability. This borrows from config updater except the collector dependencies.  Please take a look and let me know if this addresses your concern. 

https://github.com/SigNoz/signoz/pull/2418/files/0fe8932a6fc577f6b3d47f5f2301d9ea6e1e4367..56148c54dcb4f9c15aee86826c5cc27488551326#diff-470c7ea345b8f9ff994f36e2fc6216c8eec47640201f3e5cbf78eaca1fe768e6

srikanthccv replied:
>It allows flexibility in handling config and keeping up with upstream config

We are not keeping up with the upstream. There is always a version we say we support and we only work with that version because signoz-otel-collector otherwise wouldn't run. 

>I am not sure the config updater should be used to introduce the collector dependencies.

I want to understand the reluctance here. If we really want to validate the config correctly, I do not know how we can avoid it. I am not finding the idea of copying the config structs into query-service good. 

>By adding collector dep, we have to keep up the libs in sync with every dependency the collector, and its deps have. I tried doing this, and I would suggest keeping this code (even if it is not pretty) vs adding collector dep.

We are already adding dependency with "go.opentelemetry.io/collector/confmap". pretty is not a concern at all. How do we validate as the collector validates and use the structs that work with signoz-otel-collector. I ran into dep issues as well, and we are resolving them starting with `SigNoz/prometheus` it should be fine now if there is something left, it should be doable. Unless deps resolving taking a lot of effort repeatedly (not a one time) I see this as a weak argument.

srikanthccv replied:
My main concern is validating the config before sending it to the agent. If that's something we are bothered about initally, don't have any issue with `ConfigParser{}`.

mindhash replied:
The config validation can happen through dryRun as well which is infact a better indicator of success than just plainly validating config through references.

mindhash replied:
Maintaining the deps is not going to be one time effort. You would end up in dependency issues again and again. 

> Unless deps resolving taking a lot of effort repeatedly (not a one time) I see this as a weak argument

mindhash replied:
> I want to understand the reluctance here. If we really want to validate the config correctly, I do not know how we can avoid it. I am not finding the idea of copying the config structs into query-service good.

mindhash replied:
> I want to understand the reluctance here. If we really want to validate the config correctly, I do not know how we can avoid it. I am not finding the idea of copying the config structs into query-service good.

Again this is team's call. My suggestion would be not to introduce collector dependencies as it does not completely resolve config validation. The best way to know if a config would work on a collector would be to dry run it in collector itself by building pipeline. 

Adding a direct dependency will lock you in to keep up the collector's dependencies.  It goes from simple things like 'github.com/yaml3` to anything that may be commonly used on both sides. So i do not think it is a good idea but as I said this will be team's call. 

I would like to know @nityanandagohain @ankitnayan thoughts as well before adding collector dependency. 

Even if this needs to be added, I would hope this can be done in next release.

mindhash replied:
May be when opamp server is separated out from query-service, the hard dependencies can be introduced.

srikanthccv replied:
Ok summarising,

The case against doing it
- it adds dependencies and potential issues

The case for doing it
- work with the component structs directly, don't have to bother about the key names and type guarantee provided for values.
- avoid copying (and re-copying when the collect config introduces changes) the structs for the receiver, stanza parser or whatever else the code manipulates
- preliminary validation is done before sending to an agent (this is complimentary to any additional mechanism, including actually running the collector with a dry run)

Whenever we update the `signoz-otel-collector` version (ex from 0.45 to 0.66), there is most likely an unavoidable effort to get our components to work. will the dependency issue result in a blocker that we can't resolve them? Or does it take days to resolve them? then not adding dependency makes sense. Otherwise, does it take similar or less effort as collector update? Then this can't be an excuse, given the plus side of it.

mindhash commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,227 @@
+
+	_, err := r.db.ExecContext(ctx, updateQuery, status, result, lastHash, lastconf, version, string(elementType))
+	if err != nil {
+		zap.S().Error("failed to get ingestion rule from db", err)
+		return model.BadRequestStr("failed to get ingestion rule from db")
```
done

mindhash commented on pkg/query-service/agentConf/db.go:
```diff
@@ -0,0 +1,227 @@
+
+	_, err := r.db.ExecContext(ctx, updateQuery, status, result, confighash)
+	if err != nil {
+		zap.S().Error("failed to get ingestion rule from db", err)
+		return model.BadRequestStr("failed to get ingestion rule from db")
```
done

mindhash commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// UpsertSamplingProcessor updates the agent config with new filter processor params
+func UpsertSamplingProcessor(ctx context.Context, version int, config *tsp.Config) error {
```
If they are almost same I will look at re-using the function.

mindhash replied:
I am re-using the existing procs now.

mindhash commented on pkg/query-service/agentConf/manager.go:
```diff
@@ -56,7 +56,7 @@ func GetConfigHistory(ctx context.Context, typ ElementTypeDef) ([]ConfigVersion,
 
 // StartNewVersion launches a new config version for given set of elements
-func StartNewVersion(ctx context.Context, eleType ElementTypeDef, elementIds []string) (*ConfigVersion, error) {
+func StartNewVersion(ctx context.Context, userId string, eleType ElementTypeDef, elementIds []string) (*ConfigVersion, error) {
```
@nityanandagohain  this is a new change to capture user Id. The context requires more thought, hence I am opting out of it for now

nityanandagohain replied:
cool.

mindhash commented:
@srikanthccv  I have addressed most of the actionable comments except these 2:
- Before redeployment the config version status can be checked to see if it is re-deployable - **will submit a separate PR on this as this requires some additional testing  and not really breaking changes** 
- Add direct deps on otel configs - **this can be done as we learn more**  

If you don't have any additional comments please update.  I can address them in a separate PR tomorrow. I think we should move the current PR as it is blocking Nitya's testing. 

@nityanandagohain please take a look, I added minor change to StartNewVersion() by including userId in parameters. I have tested the changes with ingestion rules, so all should work on your side. But if not, let me know. 
If this breaks anything major on your side let me know. I can re-align things. I have added ConfigParser{} to maniupulate collector conf at one place. Feel free to use it now or in future releases.

mindhash commented:
@srikanthccv  I have addressed most of the actionable comments except these 2:
- Before redeployment the config version status can be checked to see if it is re-deployable - **will submit a separate PR on this as this requires some additional testing  and not really breaking changes** 
- Add direct deps on otel configs - **this can be done in separate PRs**  

If you have any additional comments please share.  I can address them in a separate PR tomorrow. I think we should move the current PR as it is blocking Nitya's testing. 

@nityanandagohain please take a look, I added minor change to StartNewVersion() by including userId in parameters. I have tested the changes with ingestion rules, so all should work on your side. But if not, let me know. 
If this breaks anything major on your side let me know. I can re-align things. I have added ConfigParser{} to maniupulate collector conf at one place. Feel free to use it now or in future releases.
